c2_utils/src/mfx_intel_device.cpp

```diff
@@ -0,0 +1,352 @@
+// Copyright (c) 2024 Intel Corporation
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in all
+// copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+// SOFTWARE.
+
+#include <fcntl.h>
+#include <stdint.h>
+#include <stdio.h>
+#include <string.h>
+#include <unistd.h>
+#include <log/log.h>
+#include <drm.h>
+#include <i915_drm.h>
+#include <virtgpu_drm.h>
+#include <xf86drm.h>
+
+#include "mfx_intel_device.h"
+
+#define ARRAY_SIZE(A) (sizeof(A) / sizeof(*(A)))
+
+#define GEN_VERSION_X10(dev)  ((dev)->graphics_version * 10 + (dev)->sub_version)
+
+#define VIRTGPU_PARAM_QUERY_DEV        11 /* Query the virtio device name. */
+#define VIRTGPU_PARAM_ALLOW_P2P        12
+
+struct intel_gpu_info {
+    int graphics_version;
+    int sub_version;
+    bool is_xelpd;
+};
+
+static int gem_param(int fd, int name)
+{
+    int v = -1; /* No param uses (yet) the sign bit, reserve it for errors */
+
+    struct drm_i915_getparam gp = {.param = name, .value = &v };
+    if (drmIoctl(fd, DRM_IOCTL_I915_GETPARAM, &gp))
+        return -1;
+
+    return v;
+}
+
+static int intel_gpu_info_from_device_id(uint16_t device_id, struct intel_gpu_info *i915)
+{
+    const uint16_t gen4_ids[] = { 0x29A2, 0x2992, 0x2982, 0x2972, 0x2A02, 0x2A12, 0x2A42,
```

Comment:
We have detailed legal device definition in VPL http://inlubt0232.iind.intel.com:8080/source/xref/CELADON/vendor/intel/external/onevpl-intel-gpu/_studio/shared/include/mfxstructures-int.h#111. We can use that for the reference.

Thanks,
Tianmi

Reply:
I think we don't need a comprehensive list of all supported devices now -- we only need to distinguish DG2. So current implementation can accomplish our purpose already.

Reply:
It is better we can keep it aligned with vpl here https://github.com/intel/vpl-gpu-rt/blob/main/_studio/shared/include/mfxstructures-int.h#L510 . If some device ids are not included or declared there, the decoding may not be supported with current vpl implementation.

Thanks,
Tianmi

Reply:
Updated to align with VPL.